### PR TITLE
Fix incorrect serialization of fixed-length arrays

### LINKED ISSUE #22: Incorrect serialization of fixed-length arrays
```
import { MessageDefinition } from "@foxglove/message-definition";
import { MessageWriter } from "@foxglove/rosmsg2-serialization";

const msgDef: MessageDefinition[] = [
  {
    definitions: [
      {
        type: "float64",
        name: "array",
        isArray: true,
        arrayLength: 10,
      },
    ],
  },
];
const msgWriter = new MessageWriter(msgDef);
const output = msgWriter.writeMessage({});
console.log(output.byteLength) // FAIL: expected 84, actual 4
```

### PR DESCRIPTION
**Pull request by achim-k**
### Public-Facing Changes

Fix incorrect serialization of fixed-length arrays

### Description
Makes sure that fixed-length arrays are default initialized if no value is given

Fixes #22 
Resolves FG-5529

### CODE REVIEW

File: src/MessageWriter.ts [hunk 1]
```diff
@@ -213,7 +232,7 @@ export class MessageWriter {
         } else {
           // Primitive array
           const arrayWriter = this.getPrimitiveArrayWriter(field.type);
-          arrayWriter(nestedMessage, field.defaultValue, writer);
+          arrayWriter(nestedMessage, field.defaultValue, writer, field.arrayLength);
```
Comment (achim-k) [inline src/MessageWriter.ts:235]: The following would have also worked without changing the array writer signature, but then typings wouldn't be correct anymore.
```suggestion
          arrayWriter(nestedMessage, field.defaultValue ?? new Array(field.arrayLength), writer);
```

**Review by jhurliman — APPROVED** [review]

File: src/MessageWriter.ts [hunk 2]
```diff
@@ -373,112 +407,162 @@ function uint8Array(value: unknown, defaultValue: DefaultValue, writer: CdrWrite
...
   }
 }
 
-function timeArray(value: unknown, _defaultValue: DefaultValue, writer: CdrWriter): void {
+function timeArray(
+  value: unknown,
+  _defaultValue: DefaultValue,
+  writer: CdrWriter,
+  _arrayLength?: number,
+): void {
```
Comment (jtbandes) [inline src/MessageWriter.ts:565]: why no defaultValue fallback here? Seems like we would need it for time arrays as well... or is there no such thing as a fixed-length time array?

File: src/MessageWriter.ts [hunk 3]
```diff
@@ -203,6 +213,15 @@ export class MessageWriter {
           writer.sequenceLength(arrayLength);
         }
 
+        if (field.arrayLength != undefined && nestedMessage != undefined) {
+          const givenFieldLength = fieldLength(nestedMessage);
+          if (givenFieldLength !== field.arrayLength) {
+            throw new Error(
+              `Given array does not match the defined array size: ${givenFieldLength} != ${field.arrayLength}`,
```
Comment (jtbandes) [inline src/MessageWriter.ts]: wording suggestion: `` `Expected ${field.arrayLength} items for fixed-length array field ${field.name} but received ${givenFieldLength}` `` ... or something like that (I think including the field name is helpful for debugging)

File: src/MessageWriter.ts [hunk 4]
```diff
@@ -203,6 +213,15 @@ export class MessageWriter {
           writer.sequenceLength(arrayLength);
         }
 
+        if (field.arrayLength != undefined && nestedMessage != undefined) {
```
Comment (jtbandes) [inline src/MessageWriter.ts:216]: Should we also validate somewhere (maybe not here) that `fieldLength(defaultValue) === arrayLength`?

(Is it even possible to have a default value for fixed-length array fields?)

File: src/MessageWriter.ts [hunk 5]
```diff
@@ -344,27 +363,42 @@ function time(value: unknown, _defaultValue: DefaultValue, writer: CdrWriter): v
...
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (Array.isArray(value)) {
     const array = new Int8Array(value);
     writer.int8Array(array);
   } else {
-    writer.int8Array((defaultValue ?? []) as number[]);
+    writer.int8Array((defaultValue ?? new Array(arrayLength).fill(0)) as number[]);
```
Comment (jtbandes) [inline src/MessageWriter.ts:376]: Just wondering, it seems like these CdrWriter `writer.*array` methods have a `writeLength` parameter... why is the MessageWriter not using that but instead doing its own `sequenceLength` call on L213?

File: src/MessageWriter.ts [hunk 6]
```diff
@@ -203,6 +213,15 @@ export class MessageWriter {
           writer.sequenceLength(arrayLength);
```
Comment (jtbandes) [inline src/MessageWriter.ts:213]: Just wondering, it seems like these CdrWriter `writer.*array` methods have a `writeLength` parameter... why is the MessageWriter not using that but instead doing its own `sequenceLength` here?

File: src/MessageWriter.ts (hunk 4 above)
Comment (achim-k) [reply src/MessageWriter.ts:216]: I don't think that's necessary. ROS2 already ensures that during the build process:

```py
# msg def: fieldtype fieldname fielddefaultvalue
int32[5] five_integers_array [0, 0, 0, 0]
```

Error when building:
```
  rosidl_adapter.parser.InvalidValue: value '[0, 0, 0, 0]' can not be
  converted to type 'int32[5]': array must have exactly 5 elements, not 4
```

File: src/MessageWriter.ts (hunk 6 above)
Comment (achim-k) [reply src/MessageWriter.ts:213]: Not sure, maybe that's something @jhurliman can answer?

Comment (jhurliman) [reply src/MessageWriter.ts:213]: I forget if there is an intentional reason or not :-/

Comment (snosenzo) [reply src/MessageWriter.ts:213]: You should definitely be able to use that parameter, it could've been added after this was implemented?

**Review by snosenzo — APPROVED** [review]

File: src/MessageWriter.ts (hunk 6 above)
Comment (achim-k) [reply src/MessageWriter.ts:213]: Not sure if it makes sense to change it in this PR. I would leave it for later

**Review by jtbandes — APPROVED** [review]

**Pull request merged**